### PR TITLE
Use stable Python C API for building Rust extension

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit tweaks the rust extension code to start using the PyO3 abi3 flag to build binaries that are compatible with all python versions, not just a single release. Previously, we were building against the version specific C API and that resulted in needing a binary file for each supported python version on each supported platform/architecture. By using the abi3 feature flag and marking the wheels as being built with the limited api we can reduce our packaging overhead to just having one wheel file per supported platform/architecture.

The only real code change needed here was to update the memory marginalization function. PyO3's abi3 feature is incompatible with returning a big int object from rust (the C API they use for that conversion isn't part of the stable C API). So this commit updates the function to convert to create a python int manually using the PyO3 api where that was being done before.

### Details and comments

TODO:

- [x] Release note
- [x] Test cibuildwheel with this change
- [x] Run Benchmarks to see if there is a performance impact

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
fwiw I'm happy with the code changes in here, but we need another reviewer (and the release note + cibuildwheel checks) because I was involved in writing the patch.

File: crates/accelerate/Cargo.toml
Comment (mtreinish) [reply crates/accelerate/Cargo.toml:29]: Yeah, I think that's best practice for this. We're already restricted to installing on python  >=3.8 so setting a minimum version in the compiled component makes sense

**Review by Eric-Arellano — APPROVED** [review]
Excellent. We were very sad when we had to stop using abi3 in Pantsbuild because of some IO stuff we were doing. This is a good change.

+1 to setting to abi3-py38

**Review by Eric-Arellano — APPROVED** [review]

**Comment by mtreinish** [comment]
Hmm, the test failures (both unit test and neko) look like there is an issue using `Complex64` and abi3 in pyo3. This will be a blocker for this because we need to be able to pass complex numbers back and forth to rust code. I guess we could refactor the code to manually pass 2 floats for the real and imaginary parts for each complex number, and build `Complex64`s in the rust from that. But that doesn't seem like a great solution

**Comment by jakelishman** [comment]
It looks like the problem here is actually a little split between us and PyO3, but fortunately enough of the bug is on our end that we should be able to get a working version.  PyO3 has (for some reason) special handling to allow conversion of length-1 arrays to scalars, but the conversion into `Complex64` via this scalarisation path doesn't appear to work correctly when it's in limited-API mode.

For example, I made a dummy PyO3 project with a Rust function:
```rust
use num_complex::Complex64;

fn noop_complex(a: Complex64) -> Complex64 {
    a
}
```
I wrapped that into a PyO3 module `pyo3_test`, and then used the Python test script:
```python
import numpy as np
from pyo3_test import noop_complex

noop_complex(np.array([1j]))
```

If you _don't_ set the limited API, the return result is `1j`.  If you do, you get
```text
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 4
      1 import numpy as np
      2 from pyo3_test import noop_complex
----> 4 noop_complex(np.array([1j]))

TypeError: argument 'a': float() argument must be a string or a real number, not 'complex'
```

That's the same error in the Neko test.  It's not identical to the one in the regular test suite, but I'm guessing/hoping it's related.

**Comment by jlapeyre** [comment]
> the problem here is actually a little split between us and PyO3 ...

"All checks have passed". But I don't see any commits address the conversion of complex problem. Am I missing something?

**Comment by mtreinish** [comment]
Oops, I didn't actually mean to commit `"auto"` I set that locally for testing and came to the same conclusion that we needed to set it explicitly. I forgot to undo my local edits before running `git commit -a` when I made the phase type fixes. Let me revert that and go back to `True`

**Review by jakelishman — DISMISSED** [review]
I'm happy with this, though we ought to have a second in-depth review.

Fwiw, I filed the PyO3 inconsistency we found (via our own bug) as PyO3/pyo3#3164.

**Review by Eric-Arellano — DISMISSED** [review]

**Review by kevinhartman — DISMISSED** [review]
LGTM once a release note is added.

**Comment by jakelishman** [comment]
We do release notes for all user-facing changes, and this is one that absolutely affects downstream packagers of Qiskit (of which there are a few).

**Comment by mtreinish** [comment]
I ran a quick comparison benchmark locally using a subset of the asv suite on 3.8, 3.9, and 3.11 (I didn't have 3.10 installed so that didn't run). It doesn't look like there will be a noticeable performance difference:

```
Benchmarks that have stayed the same:

       before           after         ratio
     [631b6f3a]       [7321041a]
     <lint_incr_latest~1>       <abi3>    
          190±2ms          200±3ms     1.05  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1) [neo-zeong/virtualenv-py3.8]
          232±1ms          240±2ms     1.03  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1) [neo-zeong/virtualenv-py3.11]
          227±1ms          234±3ms     1.03  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1) [neo-zeong/virtualenv-py3.11]
          888±9ms         914±10ms     1.03  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3) [neo-zeong/virtualenv-py3.11]
          305±4ms          314±3ms     1.03  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2) [neo-zeong/virtualenv-py3.8]
        374±0.7ms          383±6ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3) [neo-zeong/virtualenv-py3.9]
          308±2ms          315±1ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2) [neo-zeong/virtualenv-py3.8]
       1.95±0.01s       1.99±0.01s     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2) [neo-zeong/virtualenv-py3.11]
          342±2ms          349±5ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1) [neo-zeong/virtualenv-py3.8]
          376±1ms          382±3ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3) [neo-zeong/virtualenv-py3.8]
          255±2ms          259±2ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0) [neo-zeong/virtualenv-py3.11]
          257±4ms          261±5ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3) [neo-zeong/virtualenv-py3.11]
          526±4ms          533±5ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2) [neo-zeong/virtualenv-py3.8]
       1.16±0.01s          1.17±0s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1) [neo-zeong/virtualenv-py3.8]
        132±0.9ms        134±0.8ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0) [neo-zeong/virtualenv-py3.11]
          345±3ms          349±4ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0) [neo-zeong/virtualenv-py3.9]
          299±1ms          303±1ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1) [neo-zeong/virtualenv-py3.9]
          304±1ms          307±2ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0) [neo-zeong/virtualenv-py3.11]
          342±1ms          346±1ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1) [neo-zeong/virtualenv-py3.9]
          294±1ms        297±0.3ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2) [neo-zeong/virtualenv-py3.9]
          383±1ms          387±3ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3) [neo-zeong/virtualenv-py3.8]
         61.6±2ms       62.1±0.9ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0) [neo-zeong/virtualenv-py3.11]
          1.18±0s          1.19±0s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3) [neo-zeong/virtualenv-py3.8]
          303±1ms          305±1ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1) [neo-zeong/virtualenv-py3.8]
       4.53±0.01s       4.57±0.04s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3) [neo-zeong/virtualenv-py3.11]
       1.17±0.01s       1.18±0.01s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3) [neo-zeong/virtualenv-py3.9]
          420±3ms          423±2ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0) [neo-zeong/virtualenv-py3.8]
       6.00±0.01s          6.04±0s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3) [neo-zeong/virtualenv-py3.8]
          868±2ms          873±6ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1) [neo-zeong/virtualenv-py3.11]
          413±1ms          416±2ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0) [neo-zeong/virtualenv-py3.9]
          1.38±0s          1.39±0s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0) [neo-zeong/virtualenv-py3.11]
          348±5ms          349±5ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1) [neo-zeong/virtualenv-py3.8]
          219±2ms          220±2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0) [neo-zeong/virtualenv-py3.9]
        254±0.9ms          255±3ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3) [neo-zeong/virtualenv-py3.11]
       1.87±0.01s          1.87±0s     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0) [neo-zeong/virtualenv-py3.8]
          106±2ms        106±0.5ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0) [neo-zeong/virtualenv-py3.8]
          2.68±0s          2.68±0s     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2) [neo-zeong/virtualenv-py3.8]
             2565             2565     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2) [neo-zeong/virtualenv-py3.11]
          343±3ms          343±2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0) [neo-zeong/virtualenv-py3.8]
          390±5ms        389±0.9ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2) [neo-zeong/virtualenv-py3.11]
          196±2ms          196±3ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2) [neo-zeong/virtualenv-py3.11]
       1.84±0.01s       1.84±0.01s     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0) [neo-zeong/virtualenv-py3.9]
          189±5ms          189±1ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1) [neo-zeong/virtualenv-py3.9]
          1.15±0s       1.14±0.01s     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1) [neo-zeong/virtualenv-py3.9]
          337±1ms          335±2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1) [neo-zeong/virtualenv-py3.9]
          524±2ms        521±0.7ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2) [neo-zeong/virtualenv-py3.9]
       2.67±0.01s       2.66±0.01s     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2) [neo-zeong/virtualenv-py3.9]
          228±2ms        226±0.9ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0) [neo-zeong/virtualenv-py3.8]
       5.97±0.02s          5.93±0s     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3) [neo-zeong/virtualenv-py3.9]
          382±3ms          379±4ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3) [neo-zeong/virtualenv-py3.9]
          304±1ms          302±3ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2) [neo-zeong/virtualenv-py3.9]
          126±3ms          124±4ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1) [neo-zeong/virtualenv-py3.11]
          102±2ms        101±0.7ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0) [neo-zeong/virtualenv-py3.9]
          260±2ms        255±0.7ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1) [neo-zeong/virtualenv-py3.11]

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

File: crates/accelerate/src/results/marginalization.rs
Comment (jakelishman) [inline crates/accelerate/src/results/marginalization.rs:160]: Depending on when we merge this, there's a PR on PyO3 to add support for `num-bigint`+`abi3` in the works: PyO3/pyo3#3198.

**Comment by mtreinish** [comment]
This should be ready for final review now. I've added the missing release note and also tested that cibuildwheel works and is producing a single binary and testing across all python versions.

File: pyproject.toml
Comment (jakelishman) [inline pyproject.toml:31]: For posterity: the pre-`&&` components of the repair commands for Linux and macOS are copied from the `cibuildwheel` defaults.  Windows has no default command.

File: releasenotes/notes/use-abi3-4a935e0557d3833b.yaml
Comment (jakelishman) [inline releasenotes/notes/use-abi3-4a935e0557d3833b.yaml:4]:
```suggestion
    By default Qiskit builds its compiled extensions using the
```

File: releasenotes/notes/use-abi3-4a935e0557d3833b.yaml
Comment (jakelishman) [inline releasenotes/notes/use-abi3-4a935e0557d3833b.yaml:14]: Might we want to change the text about "Python 3.8" to say "the oldest version of Python supported by Qiskit (currently 3.8)", or something similar?

_Maybe_ we want to flag that we don't expect any performance regressions, too?

**Review by jakelishman — COMMENTED** [review]
Aside from the release-note comments, I'm happy to merge this at this point.

I left a couple of comments inline for posterity.  Another one Matt and I discussed more offline: it's not technically necessary to specify `abi3-py38` in the `Cargo.toml` files when passing `--py-limited-api=cp38` to `bdist_wheel` as we do universally in `setup.py` (it's inferred by `setuptools-rust`), but we chose to do it this way so that the `build_rust` command to `setup.py` and direct `cargo`-based commands will still build in limited API mode, since these are common workflows for our Rust developers.  The number of times we'll need to change the flag is very small, and once we're at an MSRV of 1.64, we can inherit one rule from the workspace's `Cargo.toml` anyway, reducing the duplication.

File: releasenotes/notes/use-abi3-4a935e0557d3833b.yaml
Comment (mtreinish) [reply releasenotes/notes/use-abi3-4a935e0557d3833b.yaml]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10120/commits/6257bdd2e749368eba8a95f8612ed556b2c13580

File: releasenotes/notes/use-abi3-4a935e0557d3833b.yaml
Comment (jakelishman) [inline releasenotes/notes/use-abi3-4a935e0557d3833b.yaml:13]:
```suggestion
    would only work with a single Python version. This change was made
```

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
I'll leave this unmerged for a little while to see if any of the other people interested in the Rust/Python API layer want to comment on the final form, since I was very involved with this PR.

**Review by Eric-Arellano — APPROVED** [review]

**Comment by jakelishman** [comment]
Merging now, since it's been a week.

**Pull request merged**
